# eBay Web Scraping Project
We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website, the data available [here](https://data.world/data-society/used-cars-data).

The aim of this project is to clean the data and analyze the used car listings.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The data contains 50,000 entries and 20 columns. Only 5 of the columns contain null-values and none contain more than 20% null. 

Most importantly for us, the column names are in *camelcase* instead of *snakecase*, the convention in Python, so we will first make the change.

## Clean Columns

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

- Change the columns from camelcase to snakecase.
- Change a few wordings to more accurately describe the columns.

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning.
Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

The `seller` and `offer_type` columns contain almost all identical values so we'll remove them. The `nr_of_pictures` column seems to have some error as well.

In [7]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

Looks as if the `nr_of_pictures` column contains no useful data, so we will remove it along with the two prior columns mentioned.

In [8]:
autos = autos.drop(["offer_type", "seller", "nr_of_pictures"], axis = 1)

The `price` and `odometer` columns need to be changed to numeric values.

In [9]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [10]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

In [11]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [12]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

Let's go ahead and update the `odometer` column to `odometer_km`.

In [13]:
autos.rename({"odometer":"odometer_km"},axis = 1, inplace = True)

## Exploring the Odometer and Price Columns


In [14]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())
autos["odometer_km"].value_counts().head(20)

(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see the number of odometer readings at 13 distinct values which leads me to believe there was a rounding by sellers likely set up through eBay. Another detail to note is nearly 80% of cars have over 100,000 miles. Overall the column looks good.

In [15]:
print(autos["price"].unique().shape)
print(autos["price"].describe().apply(lambda x: '{:.2f}'.format(x))
)
autos["price"].value_counts().head(20)

(2357,)
count       50000.00
mean         9840.04
std        481104.38
min             0.00
25%          1100.00
50%          2950.00
75%          7200.00
max      99999999.00
Name: price, dtype: object


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

We can see 1421 columns with `$0` as the price to be removed. Also the `max` price is just under $100,000,000.

In [16]:
autos["price"].value_counts().sort_index(ascending = False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [17]:
autos["price"].value_counts().sort_index(ascending = False).tail(15)

18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price, dtype: int64

Since eBay is an auction-based website we will keep for now the listings starting at `$1`. However, we will be removing all listings that are `$0` or greater than `$350,000`, as the numbers higher seem to jump substantially.

In [18]:
autos[autos["price"].between(1,350_001)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[48565 rows x 17 columns]

## Exploring the date columns
There are 5 columns that should represent date values.
- `date_crawled`: added by the crawler
- `ad_created`: from the website
- `last_seen`: added by the crawler
- `registration_month`: from the website
- `registration_year`: from the website

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.


In [19]:
autos[['date_crawled','ad_created','last_seen','registration_month','registration_year']][0:5]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_month  registration_year  
0                   3               2004  
1                   6               1997  
2                   7               2009  
3                   6               2007  
4                   7               2003

Let's try to find the distribution of values in the `date_crawled`, `ad_created`, and `last_seen columns` (all string columns) as percentages.

### `date_crawled`

In [21]:
(autos['date_crawled']
         .str[:10]
         .value_counts(normalize = True, dropna = False)
         .sort_index())

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

Let's sort instead by *values*

In [22]:
(autos['date_crawled']
         .str[:10]
         .value_counts(normalize = True, dropna = False)
         .sort_values(ascending = False))

2016-04-03    0.03868
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-12    0.03678
2016-03-14    0.03662
2016-04-04    0.03652
2016-03-07    0.03596
2016-04-02    0.03540
2016-03-19    0.03490
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-15    0.03398
2016-04-01    0.03380
2016-03-30    0.03362
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-22    0.03294
2016-03-26    0.03248
2016-03-11    0.03248
2016-03-23    0.03238
2016-03-10    0.03212
2016-03-31    0.03192
2016-03-25    0.03174
2016-03-17    0.03152
2016-03-27    0.03104
2016-03-16    0.02950
2016-03-24    0.02910
2016-03-05    0.02538
2016-03-13    0.01556
2016-03-06    0.01394
2016-04-05    0.01310
2016-03-18    0.01306
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

So the data spans from early March to early April 2016 with a fairly uniform distribution for each day (no day contains over 4% of sales)

Let's do the same now with the other columns.

##  `ad_created`

In [27]:
(autos['ad_created']
         .str[:10]
         .value_counts(normalize = True, dropna = False)
         .sort_index()
        )

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [28]:
(autos['ad_created']
         .str[:10]
         .value_counts(normalize = True, dropna = False)
         .sort_values(ascending = False)
        )

2016-04-03    0.03892
2016-03-20    0.03786
2016-03-21    0.03772
2016-04-04    0.03688
2016-03-12    0.03662
               ...   
2016-01-03    0.00002
2016-01-13    0.00002
2016-01-22    0.00002
2015-06-11    0.00002
2015-09-09    0.00002
Name: ad_created, Length: 76, dtype: float64

As we can see most of the ads were created closer to data scraping, but go as far back as the previous year.

## `last_seen`

In [30]:
(autos['last_seen']
         .str[:10]
         .value_counts(normalize = True, dropna = False)
         .sort_index(ascending = False)
        )

2016-04-07    0.13092
2016-04-06    0.22100
2016-04-05    0.12428
2016-04-04    0.02462
2016-04-03    0.02536
2016-04-02    0.02490
2016-04-01    0.02310
2016-03-31    0.02384
2016-03-30    0.02484
2016-03-29    0.02234
2016-03-28    0.02086
2016-03-27    0.01602
2016-03-26    0.01696
2016-03-25    0.01920
2016-03-24    0.01956
2016-03-23    0.01858
2016-03-22    0.02158
2016-03-21    0.02074
2016-03-20    0.02070
2016-03-19    0.01574
2016-03-18    0.00742
2016-03-17    0.02792
2016-03-16    0.01644
2016-03-15    0.01588
2016-03-14    0.01280
2016-03-13    0.00898
2016-03-12    0.02382
2016-03-11    0.01252
2016-03-10    0.01076
2016-03-09    0.00986
2016-03-08    0.00760
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
Name: last_seen, dtype: float64

In [26]:
(autos['last_seen']
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_values(ascending = False)
        )

2016-04-06    0.22100
2016-04-07    0.13092
2016-04-05    0.12428
2016-03-17    0.02792
2016-04-03    0.02536
2016-04-02    0.02490
2016-03-30    0.02484
2016-04-04    0.02462
2016-03-31    0.02384
2016-03-12    0.02382
2016-04-01    0.02310
2016-03-29    0.02234
2016-03-22    0.02158
2016-03-28    0.02086
2016-03-21    0.02074
2016-03-20    0.02070
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-23    0.01858
2016-03-26    0.01696
2016-03-16    0.01644
2016-03-27    0.01602
2016-03-15    0.01588
2016-03-19    0.01574
2016-03-14    0.01280
2016-03-11    0.01252
2016-03-10    0.01076
2016-03-09    0.00986
2016-03-13    0.00898
2016-03-08    0.00760
2016-03-18    0.00742
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
Name: last_seen, dtype: float64

The last day seen is presumably the day the car sold, so it's more likely that these spikes are to do with the crawling period ending and don't indicate car sales.

Now let's quickly take a look at the `registration_year` column.

In [32]:
autos["registration_year"].describe()


count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000 and the maximum is 9999, two values which must be mistakes.


## Dealing with Incorrect Registration Year Data
Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.


In [33]:
autos["registration_year"].between(1900,2016).value_counts(normalize=True)

True     0.96056
False    0.03944
Name: registration_year, dtype: float64

We will be removing close to 4% of the values in the table

In [34]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)


2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
2006    0.056384
2001    0.056280
2002    0.052740
1998    0.051074
2007    0.047972
Name: registration_year, dtype: float64

It appears the majority of the vehicles come from the past 20 years.

## Exploring Price by Brand
Now that the data is cleaned we can start doing some analysis. Let's first see which car brands perform the best.

In [35]:
autos["brand"].value_counts(normalize = True).head(20)

volkswagen        0.212126
bmw               0.110019
opel              0.108166
mercedes_benz     0.095361
audi              0.086387
ford              0.069793
renault           0.047347
peugeot           0.029524
fiat              0.025860
seat              0.018177
skoda             0.016032
mazda             0.015137
nissan            0.015095
citroen           0.013929
smart             0.013909
toyota            0.012472
sonstige_autos    0.010952
hyundai           0.009848
volvo             0.009245
mini              0.008641
Name: brand, dtype: float64

In [36]:
sum(autos["brand"].value_counts(normalize = True).head(6))

0.6818522528525027

We can see that the top 6 brands make up close to 70% of the total listings in the data set, so let's create a rule where only brands that account for 5% of the listings will be evaluated.

In [37]:
brand_counts = autos["brand"].value_counts(normalize = True)
common_brands = brand_counts[brand_counts > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [41]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean)

brand_mean_prices


{'volkswagen': 6516,
 'bmw': 8334,
 'opel': 5252,
 'mercedes_benz': 30317,
 'audi': 9093,
 'ford': 7263}

Mercedes is the only luxury brand amongst the best sellers in this dataset as the average prices show. It would be interesting to explore further how Mercedes compares to other luxury brands and how resale values compare for luxury vs. non-luxury vehicles.

## Storing Aggregate Data in a DataFrame
Let's now check mean mileage on the top vehicles by creating a new dataframe including both price and mileage.

In [42]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

    
brand_mean_mileage

{'volkswagen': 128730,
 'bmw': 132434,
 'opel': 129227,
 'mercedes_benz': 130860,
 'audi': 129287,
 'ford': 124046}

In [43]:
bmp_series = pd.Series(brand_mean_prices).sort_values(ascending = False)
bmm_series = pd.Series(brand_mean_mileage).sort_values(ascending = False)

In [44]:
brand_info = pd.DataFrame(bmp_series, columns = ["Mean Price"])
brand_info

Mean Price
mercedes_benz       30317
audi                 9093
bmw                  8334
ford                 7263
volkswagen           6516
opel                 5252

In [45]:
brand_info["Mean Mileage"] = bmm_series
brand_info

Mean Price  Mean Mileage
mercedes_benz       30317        130860
audi                 9093        129287
bmw                  8334        132434
ford                 7263        124046
volkswagen           6516        128730
opel                 5252        129227

The range of car mileages does not vary as much as the prices do by brand.